gpio: sam0: add interrupt support.

nzmichaelh opened the pull request:
For inputs, the chip supports highz, pull up, and pull down.

For interrupts, the chip supports single edges, both edges, and levels
with or without a filter.

Signed-off-by: Michael Hope <mlhx@google.com>

sknsean commented:
Great work
Just checked with respect to the SAMD20 👍 
looking forward to test this :-)

mnkp commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -88,23 +172,103 @@ static int gpio_sam0_read(struct device *dev, int access_op, u32_t pin,
+	/* Route and enable the interrupt.  This is safe as nothing
+	 * happens until an individual channel is unmasked.
+	 */
+	IRQ_CONNECT(EIC_IRQn, 0, gpio_sam0_isr, DEVICE_GET(gpio_sam0_0), 0);
```
We shouldn't force interrupt priority to a specific value, regardless of the fact how well we write interrupt handler. This is a sensitive subject. Let's have interrupt priority configurable via DTS (or Kconfig).

nzmichaelh replied:
Ack, will move to DTS.

nzmichaelh replied:
Done.

mnkp commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -88,23 +172,103 @@ static int gpio_sam0_read(struct device *dev, int access_op, u32_t pin,
+	/* Route and enable the interrupt.  This is safe as nothing
+	 * happens until an individual channel is unmasked.
+	 */
+	IRQ_CONNECT(EIC_IRQn, 0, gpio_sam0_isr, DEVICE_GET(gpio_sam0_0), 0);
```
We hard-code pointer to  gpio_sam0_0 (Port A). This driver will not work correctly for gpio_sam0_1 (Port B) device. An example of using IRQ_CONNECT macro with multiple devices can be found in drivers/i2c/i2c_sam_twi.c.

nzmichaelh replied:
Ack.  The EIC on the SAM0 is a bit different from other chips - it's a single 16 line unit that can be shared between different ports, but that means you need some allocation / routing logic so that if both, say, PA15 and PB15 want an interrupt then it fails, and so when an EIC interrupt fires it can turn it back to a port.

This version shortcuts that by only supporting interrupts on port A.  I'll add the routing logic.

nzmichaelh replied:
The change is turning out to be epic :)  I've pushed a WIP version that splits the EIC out into an arch specific driver so it can be used by multiple GPIO ports.

mnkp commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -4,48 +4,132 @@
+	u32_t cb_pins;
+};
+
+DEVICE_DECLARE(gpio_sam0_0);
```
This line should be placed in the device section at the bottom part of the file.

nzmichaelh replied:
Done.

mnkp commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -4,48 +4,132 @@
+
+DEVICE_DECLARE(gpio_sam0_0);
+
+#define DEV_CFG(dev)							     \
```
In Zephyr we tend to not align \ at column 80 but rather place it directly (one space distance) after the last character on the line, just as it was before the change.

nzmichaelh replied:
Ack.  I'm using clang-format for style and will update the defaults & fix.

nzmichaelh replied:
Done.

nzmichaelh commented:
Hi mnkp@ it was a mildly epic change, but PTAL.

The SAM0 has an External Interrupt Controller (EIC) unit that is shared between GPIO ports and halves of the port.  This version turns the EIC into a real device in Device Tree and then wires the GPIO ports into it.

mnkp commented on drivers/gpio/CMakeLists.txt:
```diff
@@ -18,6 +18,6 @@ zephyr_sources_ifdef(CONFIG_GPIO_QMSI       gpio_qmsi.c)
 zephyr_sources_ifdef(CONFIG_GPIO_SCH        gpio_sch.c)
 zephyr_sources_ifdef(CONFIG_GPIO_STM32      gpio_stm32.c)
-zephyr_sources_ifdef(CONFIG_GPIO_SAM0       gpio_sam0.c)
+zephyr_sources_ifdef(CONFIG_GPIO_SAM0       gpio_sam0.c gpio_sam0_eic.c)
```
I'm not an expert on DTS but it feels a bit like a hack. Shouldn't EIC - being an independent driver - have its own enable option in DTS which is then set by the GPIO driver. The end effect would be exactly the same but the design somewhat cleaner. I.e. here in CMakeList they would be enabled independently and the dependency would be defined in DTS. Especially that gpio_sam0.c is written in a way that parts of the code dependent on EIC are not compiled if EIC is not enabled.

nzmichaelh replied:
I'm happy either way.  v1 mixed the drivers, v2 (which I never pushed) had the EIC as it's own driver with an ugly API, and v3 (this version) tightly couples them.

mnkp reviewed (commented):
Thanks @nzmichaelh for the update. I'm sorry for the late reply, I'm not able to contribute too much time to Zephyr anymore. I haven't looked into EIC driver itself yet but a partial review follows.

I assume Atmel had some good reasons for designing their chip this way, still it's a good example of software unfriendly hardware design.

nzmichaelh commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -4,12 +4,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#include <errno.h>
```
Ack.  The previous includes were copy-pasted from another driver so I went through and checked what is used.  The original list was:

* kernel.h - drop as nothing used from it
* device.h - keep due to struct device
* init.h - drop as nothing used it
* errno.h - add as directly used

nzmichaelh commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -94,10 +117,10 @@ static const struct gpio_driver_api gpio_sam0_api = {
 
 /* Port A */
-#ifdef CONFIG_GPIO_SAM0_PORTA_BASE_ADDRESS
+#if CONFIG_GPIO_SAM0_PORTA_BASE_ADDRESS
```
This one is tricky.  CONFIG_GPIO_SAM0_PORTA_BASE_ADDRESS is part of the dts fixup file as in:

`#define CONFIG_GPIO_SAM0_PORTA_BASE_ADDRESS ATMEL_SAM0_GPIO_41004400_BASE_ADDRESS`

When the port is enabled in device tree, the symbol has a value.  When the port is disabled the symbol is still defined but has no value.

So the symbol is always defined so you can't use #ifdef

galak commented on drivers/gpio/gpio_sam0_eic.c:
```diff
@@ -0,0 +1,262 @@
+#include <gpio.h>
+#include <soc.h>
+
+#if CONFIG_EIC_SAM0_BASE_ADDRESS
```
when would this not be defined?

nzmichaelh replied:
I'm open to suggestions here.  If you had a minimal system and didn't want GPIO interrupts then you could drop the EIC from Device Tree and save ~1 KiB.

galak replied:
Not sure I understand, what the issue with GPIO interrupts.  Guessing this has something to do with use of EIC vs normal GPIO or something.

nzmichaelh replied:
Ah, I see.  The SAM0 has a PORT peripheral that sets the pin direction, level, pull up etc and a separate shared EIC peripheral that does edge detection and generates interrupts.  The PORT pins are muxed onto the EIC inputs, so for example PA0, PA16, PB0, and PB16 all map to EIC channel 0 and hence only one can be used at once.

nzmichaelh replied:
After thinking about it, there's no real need to support configuring out the EIC so I'll drop this #if

henrikbrixandersen commented:
@nzmichaelh Could you please rebase this and remove the pull-up/pull-down functionality, which is now covered by #6343?

henrikbrixandersen commented:
@nzmichaelh Any chance of this being picked up? :)

nzmichaelh committed:
gpio: sam0: add external interrupt support.

Add a driver and Device Tree for the shared EIC peripheral.  Wire it
into the GPIO driver.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh commented:
Rebased, but still needs work.

galak commented:
@nzmichaelh still working on this?

galak commented:
Should I close this PR out for now?

ioannisg commented:
Is this a stale PR, @nzmichaelh @galak?
